### ISSUE TITLE
Indentation in generated code is wrong following embedded block

### ISSUE THREAD

**Issue by vsajip** [bottlepy/bottle#1027]
The following template:
```
<%
def foo():
  return 'foo'
%>
<p>test</p>
<%
def bar():
  return 'bar'
%>
```
generates the following code:
```

def foo():
  return 'foo'
  
  _printlist((u'<p>test</p>\n',))
  
  def bar():
    return 'bar'
    

```
This doesn't seem right. The `_printlist` call and the `def bar():` shouldn't be indented.

**Comment by Jwink3101** [comment]
I am not at a real computer to test but I think you are forgetting (or not reading the documentation) that says all indentations are ignored and you need to end the block with

    %end

Try that and see!

**Comment by Jwink3101** [comment]
Just to follow up, there is no `%` for a block before the end. Just after any kind of block start, add an `end`

**Comment by vsajip** [comment]
Yes, but this is inside an embedded block, so you can't use % end. Hence you can do
```
% if foo:
<p>Foo</p>
% end
```
but not inside an embedded multiline block delimited by
```
<%
...
%>
```
I _have_ read the documentation, but seeing that _some_ indentation is preserved inside multiline blocks, I'm trying to understand the limitations more precisely. Note that in my second example in the OP, the indentation  is actually correctly preserved, except for the last line.

**Comment by defnull** [comment]
You always have to close blocks with `end`. If the documentation is not clear enough, I'd be happy to accept pull requests.

**Comment by ralphwetzel** [comment]
To conclude: You can't use `% end` ... but you **can** use `end` inside an embedded block:

```
<%
def foo():
  return 'foo'
  end
%>
<p>test</p>
<%
def bar():
  return 'bar'
  end
%>
```

This returns:

```
def foo():
  return 'foo'
_printlist(('<p>test</p>\n',))
def bar():
  return 'bar'
```

**Comment by defnull** [comment]
> I have read the documentation, but seeing that some indentation is preserved inside multiline blocks, I'm trying to understand the limitations more precisely. Note that in my second example in the OP, the indentation is actually correctly preserved, except for the last line.

The indentation is not preserved, but recreated by bottle while parsing the template. The template engine contains a very simple modified python parser. It detects the `if foo:` lines a block start and will indent everything up until the next `end` (with special rules for `else`, `expect` and other block continuation keywords).

Try this:

    <%
         def foo():
      return 'foo'
    end
    %>

This will produce valid python code just fine :)

**Comment by defnull** [comment]
> To conclude: You can't use `% end` ... but you can use `end` inside an embedded block

Yes, exactly. The template engine does not scan for `% end` directly, it scans for the `end` keyword in code sections and these may be introduced by `%` or by a `<%%>` block (or a mixture thereof).

Fun fact: The engine is smart enough to not detect `end` within multi-line strings or as part of an expression. The following template is completely valid:

    <%
    def foo():
        end = """
        end
        """
        return end; end
   %>

Only the very last end will be used as a block-end keyword.
You can also mix `%` and `<% ... %>` if you want (not recommended, but works):

    % def foo():
    <% return "end"; end %>

**Comment by vsajip** [comment]
OK, thanks. I hadn't realised you could put `end` inside a `<% %>` multiline block. I will play around some more with `end` inside the block and either come back with more comments/questions or close the issue.
